### PR TITLE
✨ Payments service: http clients

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

This PR introduces modules to deal with the  http-api connections with RUT and the payment-gateway. Both modules have been created using common tools in `utils/http_client.py` (will eventually be moved to `servicelib.fastapi`.

payments-copy drawio

- ✨ httpx connection to RUT api
- ✨ httpx connection to payments-gateway
    - ♻️ updates openapi specs
    - tested app-motion prototype against openapi specs (SEE repo https://github.com/ITISFoundation/appmotion-exchange)
- ✨ Extend CLI`simcore-service-payments echo-dotenv  --minimal` 
   
    
## Related issue/s

- Follows up from https://github.com/ITISFoundation/osparc-simcore/pull/4715
- Implementing part of  https://github.com/ITISFoundation/osparc-simcore/issues/4657

## How to test 

- driving tests `services/payments/tests/unit/test_services_resource_usage_tracker.py`

## DevOps

- ✨ New env-vars consumed in the `payments` service ( already added in `.env-devel`)
``` 
     - RESOURCE_USAGE_TRACKER_HOST=${RESOURCE_USAGE_TRACKER_HOST}
     - PAYMENTS_SWAGGER_API_DOC_ENABLED=${PAYMENTS_SWAGGER_API_DOC_ENABLED}
``` 
- ✨ New CLI: `simcore-service-payments echo-dotenv` prints a valid dot-envfile (see --help for different options!). Example of usage:
```cmd
    $ simcore-service-payments echo-dotenv > .env
    $ cat .env
    $ set -o allexport; source .env; set +o allexport
```
NOTE: this might be useful to generate an initial list of env-vars for a services.

### CODE REVIEW

File: services/payments/src/simcore_service_payments/services/resource_usage_tracker.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+
+def setup_resource_usage_tracker(app: FastAPI):
+    assert app.state  # nosec
+    settings: ApplicationSettings = app.state.settings
+    api = ResourceUsageTrackerApi.from_client_kwargs(
+        base_url=settings.PAYMENTS_RESOURCE_USAGE_TRACKER.base_url,
+    )
+    api.set_to_app_state(app)
+    api.attach_lifespan_to(app)
```
Comment (GitHK) [inline services/payments/src/simcore_service_payments/services/resource_usage_tracker.py:74]: Q: I see a new pattern introduced here. It's something different from what we have before. I wonder if it was really that necessary to add it?

**Review by bisgaard-itis — APPROVED** [review]
LGTM

File: services/payments/src/simcore_service_payments/core/application.py [hunk 2]
```diff
@@ -13,22 +13,26 @@
...
+
+    app_settings = settings or ApplicationSettings.create_from_envs()
+
     app = FastAPI(
         title=f"{PROJECT_NAME} web API",
         description=SUMMARY,
         version=API_VERSION,
         openapi_url=f"/api/{API_VTAG}/openapi.json",
-        docs_url="/doc",  # NOTE: token auth would not work otherwise
+        docs_url="/doc" if app_settings.PAYMENTS_SWAGGER_API_DOC_ENABLED else None,
```
[the remark this answers is not in the document]
Comment (pcrespov) [reply services/payments/src/simcore_service_payments/core/application.py:29]: @bisgaard-itis this is an open entrypoint point is for the payment service and will be exposed to the internet (with auth layer, of course). nonetheless ops cannot guarantee they can auth this site.

File: services/payments/scripts/fake_payment_gateway.py [hunk 3]
```diff
@@ -60,7 +77,7 @@ def cancel_payment(
     return router
 
 
-def auth_session(x_init_api_secret: Annotated[str | None, Header()] = None):
+def auth_session(X_Init_Api_Secret: Annotated[str | None, Header()] = None):
```
[the remark this answers is not in the document]
Comment (pcrespov) [reply services/payments/scripts/fake_payment_gateway.py:80]: I was trying something ... to checka problem on app-motion service side.
fastapi takes the param name and converts it to a header-name. Headers are case-insensitive by the standard so from the point of view of the http-api is equivalent. I will change it...

File: services/payments/src/simcore_service_payments/core/application.py [hunk 4]
```diff
@@ -41,6 +45,9 @@ def create_app(settings: ApplicationSettings | None = None) -> FastAPI:
     setup_payments_gateway(app)
     setup_rest_api_routes(app)
 
+    # APIs w/ RUT
```
[the remark this answers is not in the document]
Comment (pcrespov) [reply services/payments/src/simcore_service_payments/core/application.py:48]: but it is the title for  `setup_resource_usage_tracker` ...

File: services/payments/src/simcore_service_payments/services/payments_gateway.py [hunk 5]
```diff
@@ -26,68 +23,49 @@
...
-            base_url=settings.PAYMENTS_GATEWAY_URL,
-        )
-        exit_stack = contextlib.AsyncExitStack()
-
-        return cls(client=client, exit_stack=exit_stack)
-
-    async def start(self):
-        await self.exit_stack.enter_async_context(self.client)
+    def auth_flow(self, request):
+        request.headers["X-Init-Api-Secret"] = self.token
```
[the remark this answers is not in the document]
Comment (pcrespov) [reply services/payments/src/simcore_service_payments/services/payments_gateway.py:36]: Yes I am aware. I answered above.

File: services/payments/src/simcore_service_payments/services/payments_gateway.py [hunk 6]
```diff
@@ -26,68 +23,49 @@
...
+    # api: one-time-payment workflow
     #
 
     async def init_payment(self, payment: InitPayment) -> PaymentInitiated:
-        response = await self.client.post("/init", json=jsonable_encoder(payment))
+        response = await self.client.post(
+            "/init",
+            json=jsonable_encoder(payment),
+        )
+        # FIXME: handle exceptions! SEE ErrorModel
```
[the remark this answers is not in the document]
Comment (pcrespov) [reply services/payments/src/simcore_service_payments/services/payments_gateway.py:52]: very sorry for that... error handling will come in next PR

File: services/payments/tests/unit/test_services_resource_usage_tracker.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+) -> Iterator[MockRouter]:
+    settings: ApplicationSettings = app.state.settings
+    with respx.mock(
+        base_url=settings.PAYMENTS_RESOURCE_USAGE_TRACKER.base_url,
+        assert_all_called=False,
+        assert_all_mocked=True,  # IMPORTANT: KEEP always True!
+    ) as respx_mock:
+        #
+        # /
+        #
```
[the remark this answers is not in the document]
Comment (pcrespov) [reply services/payments/tests/unit/test_services_resource_usage_tracker.py:85]: hahaha ... i get it its minimalistic but it actually means that I am mocking the route with path `/` :-D

File: services/payments/src/simcore_service_payments/services/resource_usage_tracker.py (hunk 1 above)
Comment (pcrespov) [reply services/payments/src/simcore_service_payments/services/resource_usage_tracker.py:74]: Yes, that is my point in the description. I have created some tooling to create uniform wrappers for api-clients based on httpx. The tooling incorporate the ability to inject the api as part of the ``app.state`` and attach its lifespan to the ``app`` as well.
I expect to move this to `servicelib.fastapi` when you all agree with the pattern.

**Review by GitHK — APPROVED** [review]
👍

File: services/payments/tests/unit/test_services_payments_gateway.py [hunk 8]
```diff
@@ -60,15 +97,19 @@ async def test_payment_gateway_responsiveness(
     assert await payment_gateway_api.is_healhy()
 
 
+@pytest.mark.testit
```
Comment (matusdrobuliak66) [inline services/payments/tests/unit/test_services_payments_gateway.py:100]: forgotten `@pytest.mark.testit`

**Review by matusdrobuliak66 — APPROVED** [review]
Very nice! :tada:

File: services/payments/tests/unit/test_services_payments_gateway.py (hunk 8 above)
Comment (pcrespov) [reply services/payments/tests/unit/test_services_payments_gateway.py:100]: I will remove in next PR .noe everything is green :-)

Comment (matusdrobuliak66) [reply services/payments/tests/unit/test_services_payments_gateway.py:100]: ye np :) i would recommend to force merge :+1:

**Review by YuryHrytsuk — APPROVED** [review]
Thank you!

**Comment by mrnicegyu11** [comment]
https://github.com/ITISFoundation/osparc-ops-environments/pull/386

**Review by mrnicegyu11 — APPROVED** [review]

**Pull request merged**